### PR TITLE
removed overzealous zombie process clean up and switched to dumb-init

### PR DESCRIPTION
**Pull request by GregHanson**
the current zombie process cleanup was receiving SIGCHLD's for commands started in other parts of the sidecar code (see issue #401 ).  Switched over to use dumb-init as PID to handle zombie process cleanup

### CODE REVIEW

**Comment by ijsnellf** [comment]
We'll need to update the docs as well if we merge this change.

**Review by ijsnellf — APPROVED** [review]
LGTM, but it'd be ideal to get input from @zcahana before we merge it.

File: docker/Dockerfile.sidecar.alpine [hunk 1]
```diff
@@ -132,6 +36,9 @@ RUN set -ex \
     && mkdir /lib64 \
     && ln -s /lib/libc.musl-x86_64.so.1 /lib64/ld-linux-x86-64.so.2
 
+RUN apk --no-cache add openssl
+RUN wget -O /usr/bin/dumb-init https://github.com/Yelp/dumb-init/releases/download/v1.2.0/dumb-init_1.2.0_amd64
+RUN chmod +x /usr/bin/dumb-init
```
Comment (rshriram) [inline docker/Dockerfile.sidecar.alpine:42]: is the openssl package needed ? openresty:alpine already comes with openSSL latest version

Comment (GregHanson) [reply docker/Dockerfile.sidecar.alpine:42]: it apparently does not because the alpine image fails to build without adding openssl

Comment (zcahana) [reply docker/Dockerfile.sidecar.alpine:42]: Can we do this in a single RUN directive so that we get one layer?

File: docker/Dockerfile.sidecar.ubuntu [hunk 2]
```diff
@@ -32,13 +32,16 @@ RUN set -ex \
     \
     && rm -rf /tmp/filebeat*
 
+RUN curl -sSL -o /usr/bin/dumb-init https://github.com/Yelp/dumb-init/releases/download/v1.2.0/dumb-init_1.2.0_amd64
+RUN chmod +x /usr/bin/dumb-init
```
Comment (zcahana) [inline docker/Dockerfile.sidecar.ubuntu]: ditto

File: scripts/a8sidecar.sh [hunk 3]
```diff
@@ -63,3 +63,10 @@ tar -xzf /tmp/a8sidecar-${A8SIDECAR_RELEASE}-linux-amd64.tar.gz -C /
 #Cleanup
 rm -rf ${A8TMP}
 rm /tmp/a8sidecar-${A8SIDECAR_RELEASE}-linux-amd64.tar.gz
+
+if [ $HAVE_WGET -eq 1 ]; then
```
Comment (zcahana) [inline scripts/a8sidecar.sh:67]: I suggest we not including dumb-init in our installation script, as it is _not_ a dependency for the amalgam8 sidecar.
The sidecar can perfectly operate without it, because it properly reaps its child processes, and so does nginx.

However, given that using a proper init process is considered best-practices when running multiple processes in a container, we should install and use it our prebuilt sidecar images, and probably add a doc recommendation so that users who build their own images will install dumb-init/runit/whatever, and refer them to our Dockerfile as an example.

File: testing/apps/details/Dockerfile.sidecar [hunk 4]
```diff
@@ -19,4 +19,4 @@ COPY details /opt/microservices/
 
 EXPOSE 9080
 
-ENTRYPOINT ["a8sidecar", "/opt/microservices/details", "9080" ]
\ No newline at end of file
+ENTRYPOINT ["/usr/bin/dumb-init", "--", "a8sidecar", "/opt/microservices/details", "9080" ]
```
Comment (zcahana) [inline testing/apps/details/Dockerfile.sidecar]: This is not directly related to this PR, but the bookinfo/helloworld Dockerfiles shouldn't redefine the ENTRYPOINT, but rather inherit it from the base image, and add their arguments as CMD:

```
CMD [ "/opt/microservices/details", "9080" ]
```

**Review by zcahana — CHANGES_REQUESTED** [review]
Overall looks ok, just a few little comments.

**Comment by zcahana** [comment]
@GregHanson @ijsnellf are you guys still working this? We need to get issue #401 resolved as its breaks the `master` branch for now.

One suggestion was to keep the zombie reaping logic code around, but use it only if running as PID 1, and make sure to spawn a child process for the all other functions of the sidecar. This will also require adding code to forward any signals down to the child sidecar process. I myself am ok with both approaches (dumb-init or are own zombie reaper).

**Comment by GregHanson** [comment]
Closing.  Resolved with pull request #421.
